services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_aggregate_query.rb

```diff
@@ -0,0 +1,180 @@
+# frozen_string_literal: true
+
+module AdminDiagnosticReports
+  class DiagnosticAggregateQuery < Snapshots::PeriodQuery
+    class InvalidAggregationError < StandardError; end
+
+    attr_reader :additional_aggregation
+
+    AGGREGATION_OPTIONS = [
+      'grade',
+      'teacher',
+      'classroom'
+    ]
+    DIAGNOSTIC_ORDER_BY_ID = [
+      1663, # Starter Baseline Diagnostic (Pre)
+      1668, # Intermediate Baseline Diagnostic (Pre)
+      1678, # Advanced Baseline Diagnostic (Pre)
+      1161, # ELL Starter Baseline Diagnostic (Pre)
+      1568, # ELL Intermediate Baseline Diagnostic (Pre)
+      1590, # ELL Advanced Baseline Diagnostic (Pre)
+      992,  # AP Writing Skills Survey
+      1229, #  Pre-AP Writing Skills Survey 1
+      1230, # Pre-AP Writing Skills Survey 2
+      1432  # SpringBoard Writing Skills Survey
+    ]
+
+    def initialize(aggregation:, **options)
+      raise InvalidAggregationError, "#{aggregation} is not a valid aggregation value." unless AGGREGATION_OPTIONS.include?(aggregation)
+
+      @additional_aggregation = aggregation
+
+      super(**options)
+    end
+
+    def run
+      post_process(run_query)
+    end
+
+    def select_clause
+      <<-SQL
+        SELECT
+          activities.id AS diagnostic_id,
+          activities.name AS diagnostic_name,
+          #{aggregate_by_clause} AS aggregate_id,
+          #{aggregate_sort_clause} AS name,
+          #{specific_select_clause}
+      SQL
+    end
+
+    def from_and_join_clauses
+      super + <<-SQL
+        JOIN lms.classroom_units
+          ON classrooms.id = classroom_units.classroom_id
+        JOIN lms.users
+          ON schools_users.user_id = users.id
+      SQL
+    end
+
+    def specific_select_clause
+      raise NotImplementedError
+    end
+
+    def where_clause
+      super + <<-SQL
+          #{relevant_diagnostic_where_clause}
+      SQL
+    end
+
+    def relevant_diagnostic_where_clause
+      "AND activities.id IN (#{DIAGNOSTIC_ORDER_BY_ID.join(',')})"
+    end
+
+    def group_by_clause
+      "GROUP BY activities.id, activities.name, aggregate_id, #{aggregate_sort_clause}"
+    end
+
+    def aggregate_by_clause
+      {
+        'grade' => "classrooms.grade",
+        'classroom' => "classrooms.id",
+        'teacher' => "users.id"
+      }.fetch(additional_aggregation)
+    end
+
+    def aggregate_sort_clause
+      {
+        'grade' => "classrooms.grade",
+        'classroom' => "classrooms.name",
+        'teacher' => "users.name"
+      }.fetch(additional_aggregation)
+    end
+
+    private def post_process(result)
+      return [] if result.empty?
+
+      result.group_by { |row| row[:diagnostic_name] }
```

Comment:
Looking at the amount of post-processing, I'm curious if there is a better query to run to get the data in the right shape, e.g. using a few WITH queries.